### PR TITLE
extract renderers

### PR DESCRIPTION
**Pull request by pq**
Follow-up from: https://github.com/flutter/flutter-intellij/pull/2413#issuecomment-400318578

/cc @quangson91 @scheglov

### CODE REVIEW

**Comment by pq** [comment]
@quangson91 : are you able to "review" this?  If so, could you add one?

https://help.github.com/articles/about-pull-request-reviews/

Thanks for the quick feedback! 👍

**Comment by quangson91** [comment]
@pq, Yeap. I'm checking this PR. ^^

File: src/io/flutter/logging/FlutterLogTree.java [hunk 1]
```diff
@@ -68,6 +68,68 @@ void appendStyled(FlutterLogEntry entry, String text) {
...
+    abstract class Column extends ColumnInfo<DefaultMutableTreeNode, FlutterLogEntry> {
+      boolean show = true;
+      private TableCellRenderer renderer;
+
+      Column(String name) {
+        super(name);
+      }
+
+      boolean isVisible() {
+        return show;
```
Comment (quangson91) [inline src/io/flutter/logging/FlutterLogTree.java]: `show` vs `isVisible`
-> I think it should be same.
I mean, both should be `boolean show` && `boolean isShow()`
OR
`boolean isVisible` && `boolean isVisible()`

File: src/io/flutter/logging/FlutterLogTree.java [hunk 2]
```diff
@@ -68,6 +68,68 @@ void appendStyled(FlutterLogEntry entry, String text) {
       abstract void render(FlutterLogEntry entry);
     }
 
+    abstract class Column extends ColumnInfo<DefaultMutableTreeNode, FlutterLogEntry> {
+      boolean show = true;
+      private TableCellRenderer renderer;
```
Comment (quangson91) [inline src/io/flutter/logging/FlutterLogTree.java]: add `@Nullable` maybe better for IDE lining.

File: src/io/flutter/logging/FlutterLogTree.java [hunk 3]
```diff
@@ -68,6 +68,68 @@ void appendStyled(FlutterLogEntry entry, String text) {
...
+
+      Column(String name) {
```
Comment (quangson91) [inline src/io/flutter/logging/FlutterLogTree.java]: `@NotNull`

**Comment by pq** [comment]
Cool.  Let's see if we can make it more formal.  👍 

@quangson91, when you go to:

https://github.com/flutter/flutter-intellij/pull/2430/files

do you see a review changes button in the top right?

image

File: src/io/flutter/logging/FlutterLogTree.java [hunk 4]
```diff
@@ -68,6 +68,68 @@ void appendStyled(FlutterLogEntry entry, String text) {
...
+
+      @Override
+      public TableCellRenderer getCustomizedRenderer(DefaultMutableTreeNode o, TableCellRenderer renderer) {
+        if (renderer == null) {
+          renderer = createRenderer();
+        }
+        return renderer;
+      }
+
+      abstract TableCellRenderer createRenderer();
```
Comment (quangson91) [inline src/io/flutter/logging/FlutterLogTree.java]: Nullable OR NotNull
(So sub-class will know this method can be return null or not)

File: src/io/flutter/logging/FlutterLogTree.java [hunk 5]
```diff
@@ -68,6 +68,68 @@ void appendStyled(FlutterLogEntry entry, String text) {
...
+        }
+        return null;
+      }
+
+      @Override
+      public TableCellRenderer getCustomizedRenderer(DefaultMutableTreeNode o, TableCellRenderer renderer) {
+        if (renderer == null) {
+          renderer = createRenderer();
+        }
+        return renderer;
```
Comment (quangson91) [inline src/io/flutter/logging/FlutterLogTree.java]: For me, I think we can make `render` as a `final` variable then we will init it in the constructor.
```java
      @NotNull
      private final TableCellRenderer renderer;

      Column(String name) {
        super(name);
        renderer = createRenderer();
      }
```

So in this method we can simply
```java
      @Override
      public TableCellRenderer getCustomizedRenderer(DefaultMutableTreeNode o, TableCellRenderer renderer) {
        return renderer;
      }
```

File: src/io/flutter/logging/FlutterLogTree.java [hunk 6]
```diff
@@ -157,47 +187,25 @@ public TableCellRenderer getCustomizedRenderer(DefaultMutableTreeNode o, TableCe
       columns.add(new Column("Time") {
         @Override
         TableCellRenderer createRenderer() {
-          return new EntryCellRenderer() {
-            @Override
-            void render(FlutterLogEntry entry) {
-              appendStyled(entry, TIMESTAMP_FORMAT.format(entry.getTimestamp()));
-            }
-          };
+          return new TimeCellRenderer();
```
Comment (quangson91) [inline src/io/flutter/logging/FlutterLogTree.java]: Hmn, It looks verbose.
If we custom `XXXCellRender` => I think we could do it more simply by passing `cellRender` as a parameter in constructor.
```java
    class Column extends ColumnInfo<DefaultMutableTreeNode, FlutterLogEntry> {
      boolean show = true;
      @NotNull
      private final TableCellRenderer renderer;

      Column(String name, @NotNull TableCellRenderer renderer) {
        super(name);
        this.renderer = renderer;
      }
```

So the code add columns will become:

```java
      columns.add(new Column("Time", new TimeCellRenderer()));
      columns.add(new Column("Sequence", new SequenceCellRenderer()));
      columns.add(new Column("Level", new LevelCellRenderer()));
      columns.add(new Column("Category", new CategoryCellRenderer()));
      columns.add(new Column("Message", new MessageCellRenderer(app.getModule()));
```

OR
```java
      columns.addAll(Arrays.asList(
        new Column("Time", new TimeCellRenderer()),
        new Column("Sequence", new SequenceCellRenderer()),
        new Column("Level", new LevelCellRenderer()),
        new Column("Category", new CategoryCellRenderer()),
        new Column("Message", new MessageCellRenderer(app.getModule()))
      ));
```

**Comment by quangson91** [comment]
@pq, Yeap, I can review the PR ^^.

Thank you!. I just added some comments.

**Comment by quangson91** [comment]
Just notes:
After working in this class many times, I think this class becomes bigger.
So we should consider extracting `Column/CellRender` to a single class. (then make it as `package access`).

File: src/io/flutter/logging/FlutterLogTree.java (hunk 1 above)
Comment (pq) [reply src/io/flutter/logging/FlutterLogTree.java]: Agreed!

File: src/io/flutter/logging/FlutterLogTree.java (hunk 3 above)
Comment (pq) [reply src/io/flutter/logging/FlutterLogTree.java]: 👍

File: src/io/flutter/logging/FlutterLogTree.java (hunk 4 above)
Comment (pq) [reply src/io/flutter/logging/FlutterLogTree.java]: 👍

File: src/io/flutter/logging/FlutterLogTree.java (hunk 5 above)
Comment (pq) [reply src/io/flutter/logging/FlutterLogTree.java]: There's something clumsy here I agree.  I have some thoughts about this but let's follow-up on a bigger round of cleanup.

File: src/io/flutter/logging/FlutterLogTree.java (hunk 6 above)
Comment (pq) [reply src/io/flutter/logging/FlutterLogTree.java]: Nice!  Agreed.

**Comment by pq** [comment]
> After working in this class many times, I think this class becomes bigger.

Yeah.   This is a common problem.

> So we should consider extracting Column/CellRender to a single class.

I'll take a crack at this.  I agree it could be easier to work on down the road.

Thanks!

File: src/io/flutter/logging/FlutterLogTree.java (hunk 2 above)
Comment (pq) [reply src/io/flutter/logging/FlutterLogTree.java]: 👍

File: src/io/flutter/logging/FlutterLogTree.java [hunk 7]
```diff
@@ -68,6 +68,65 @@ void appendStyled(FlutterLogEntry entry, String text) {
...
+      public FlutterLogEntry valueOf(DefaultMutableTreeNode node) {
+        if (node instanceof FlutterEventNode) {
+          return ((FlutterEventNode)node).entry;
+        }
+        return null;
+      }
+
+      @Override
+      public TableCellRenderer getCustomizedRenderer(DefaultMutableTreeNode o, TableCellRenderer renderer) {
+        return renderer;
```
Comment (quangson91) [inline src/io/flutter/logging/FlutterLogTree.java]: We should return `this. renderer`
(Because method's parameters also named renderer)

Comment (pq) [reply src/io/flutter/logging/FlutterLogTree.java]: 👍

**Pull request merged**